certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -250,6 +259,121 @@ def _get_ranked_matches(self, target_name):
                                 'rank': 6 if vhost.ssl else 7})
         return sorted(matches, key=lambda x: x['rank'])
 
+    def all_vhost_ports(self, domain):
+        """Get all ports that all vhosts are listening on, that match domain.
+
+        :param str domain: domain to find ports of
+        :returns: set of strings representing the port numbers
+        :rtype: set
+        """
+        matching_vhosts = [item['vhost'] for item in self._get_ranked_matches(domain)]
+        out = set([addr.get_port() for vhost in matching_vhosts for addr in vhost.addrs])
+        # also check the default port, 80
+        out.add(self.DEFAULT_PORT)
+        if "" in out:
+            out.remove("")
+        if r"*" in out:
+            out.remove(r"*")
+        return out
+
+    def choose_redirect_vhost(self, target_name, port):
+        r"""Chooses a virtual host for redirect enhancement.
+
+        Filter all hosts for those that include listen port but do not include
+        ssl on. Of these, find the most matching one based on target_name.
+
+        .. todo:: This should maybe return list if no obvious answer
+            is presented.
+
+        .. todo:: The special name "$hostname" corresponds to the machine's
+            hostname. Currently we just ignore this.
+
+        :param str target_name: domain name
+        :param str port: port number or "\*" or ""
+        :returns: vhost associated with name
+        :rtype: :class:`~certbot_nginx.obj.VirtualHost`
+
+        """
+        vhost = None
+
+        matches = self._get_redirect_ranked_matches(target_name, port)
+        if not matches:
+            # No matches for this port. That's fine.
+            return None
+        elif matches[0]['rank'] in xrange(2, 6):
+            # Wildcard match - need to find the longest one
+            rank = matches[0]['rank']
+            wildcards = [x for x in matches if x['rank'] == rank]
+            vhost = max(wildcards, key=lambda x: len(x['name']))['vhost']
+        else:
```

Comment:
Stylistic nit: it might be cleaner if this case came first,
```
if matches[0]['rank'] == 0:    # exactly named server block
    vhost = matches[0]['vhost']   
else:
    # wildcard or regex case
```